plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java

```diff
@@ -179,52 +180,34 @@ public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, Connect
                 .useSnapshot(snapshotId.get())
                 .includeColumnStats();
         // TODO make the cursor lazy
-        return buildRecordCursor(getPartitions(tableScan), icebergTable.spec().fields());
+        return buildRecordCursor(getStatisticsByPartition(tableScan), icebergTable.spec().fields());
     }
 
-    private Map<StructLikeWrapper, Partition> getPartitions(TableScan tableScan)
+    private Map<StructLikeWrapper, IcebergStatistics> getStatisticsByPartition(TableScan tableScan)
     {
         try (CloseableIterable<FileScanTask> fileScanTasks = tableScan.planFiles()) {
-            Map<StructLikeWrapper, Partition> partitions = new HashMap<>();
-
+            Map<StructLikeWrapper, IcebergStatistics.IcebergStatisticsBuilder> partitions = new HashMap<>();
             for (FileScanTask fileScanTask : fileScanTasks) {
                 DataFile dataFile = fileScanTask.file();
                 Types.StructType structType = fileScanTask.spec().partitionType();
                 StructLike partitionStruct = dataFile.partition();
                 StructLikeWrapper partitionWrapper = StructLikeWrapper.forType(structType).set(partitionStruct);
 
-                if (!partitions.containsKey(partitionWrapper)) {
-                    Partition partition = new Partition(
-                            idToTypeMapping,
-                            nonPartitionPrimitiveColumns,
-                            partitionStruct,
-                            dataFile.recordCount(),
-                            dataFile.fileSizeInBytes(),
-                            convertBounds(idToTypeMapping, dataFile.lowerBounds()),
-                            convertBounds(idToTypeMapping, dataFile.upperBounds()),
-                            dataFile.nullValueCounts(),
-                            dataFile.columnSizes());
-                    partitions.put(partitionWrapper, partition);
-                    continue;
-                }
-
-                Partition partition = partitions.get(partitionWrapper);
-                partition.incrementFileCount();
-                partition.incrementRecordCount(dataFile.recordCount());
-                partition.incrementSize(dataFile.fileSizeInBytes());
-                partition.updateMin(convertBounds(idToTypeMapping, dataFile.lowerBounds()), dataFile.nullValueCounts(), dataFile.recordCount());
-                partition.updateMax(convertBounds(idToTypeMapping, dataFile.upperBounds()), dataFile.nullValueCounts(), dataFile.recordCount());
-                partition.updateNullCount(dataFile.nullValueCounts());
+                partitions.putIfAbsent(
+                        partitionWrapper,
+                        new IcebergStatistics.IcebergStatisticsBuilder(idToTypeMapping, icebergTable.schema().columns(), typeManager));
+                partitions.get(partitionWrapper).registerDataFile(dataFile, fileScanTask.spec());
```

Comment:
computeIfAbsent (without another `get`) gives you 0.5x map access and skips IcebergStatisticsBuilder allocation.